guardian/support-service-lambdas
# Create a new term for subscriptions during product switch

## What does this change?

We have noticed that the current product switch from recurring contribution to supporter plus has an issue, which is that because the charge is aligned to the subscription end date, unless the switch happens on a term start date the billing amount and date will be unpredictable.

<img width="691" alt="Screenshot 2023-11-01 at 10 25 33" src="https://github.com/guardian/support-service-lambdas/assets/181371/b40ae88f-ca2c-49ae-a6fa-b6d192920bf5">

This PR updates the switch mechanism fairly substantially to avoid this, changes are:
- We check the term start date of the subscription to be switched before we begin - if it is not today we also reduce the length of the current term to today and then renew the sub for a new 12 month term. For subs where the term start date is today we skip this step.
- I have removed the requirement for the client to pass in a `checkChargeAmountBeforeUpdate` parameter which then caused us to run an additional preview call to check whether to collect the amount payable or write it off if it was below $0.50. We now just check the balance of the invoice created by the switch for this amount.
- I have moved the collection of payment to a separate api call so that we can separate out the logic involved with it (whether to collect or write off the amount owing) from the updates to the subscription that are being carried out.


## Test Subscriptions
Subscription switched on the day it was taken out - no renewal required:
https://apisandbox.zuora.com/platform/subscriptions/8ad093fb8bc82179018bc86b11d25702

Subscription switched part way through the term - renewal required:
https://apisandbox.zuora.com/platform/subscriptions/8ad097048bc82163018bc86dba57092c


## How to test
- Find a recurring contribution
- Switch it to a Supporter Plus subscription using 
```
curl --location 'https://zvobga46la.execute-api.eu-west-1.amazonaws.com/CODE/product-move/[MY-SUBSCRIPTION-NUMBER]' \
  --header 'x-api-key: [MY-API-KEY]' \
--header 'Content-Type: application/json' \
--data '{
    "price": "120",
    "preview": false,
    "checkChargeAmountBeforeUpdate": true
}'
```
- Check that the subscription has been switched to a Supporter Plus sub, that the term start date is today and the term end date is in 12 months time and that the outstanding balance on the account is zero.
